### PR TITLE
Fix/async call for mention suggestion

### PR DESCRIPTION
**Pull request by qfrank**
### Summary

This PR addresses an issue with asynchronous calls in the mention functionality. The main changes include:

1. Enhanced mention result handling:
   - Added `call-id` and `call-time` to track individual calls
   - Updated the `transfer-mention-result` function to include these new fields

2. Improved `on-change-text` event:
   - Introduced a `call-id` to uniquely identify each call
   - Added the current timestamp to the RPC call parameters

3. Refined `on-change-text-success` event:
   - Implemented a check to ensure the response matches the most recent call
   - Added response time logging for performance monitoring
   - Only update the app state if the response is for the latest call

These changes aim to resolve race conditions and ensure that the UI always reflects the most recent user input when dealing with mentions. The addition of call tracking and response time logging will help with future performance optimizations.

Relate status-go [PR](https://github.com/status-im/status-go/pull/3806)

### Testing notes
mention suggestion should work as before

#### Platforms

- Android
- iOS

status: ready

### CODE REVIEW

**Comment by flexsurfer** [comment]
thank you @qfrank , there is a PR with composer simplification, https://github.com/status-im/status-mobile/pull/20125 , there were changes and improvements for mentions as well, i would wait for it merged first

File: src/legacy/status_im/chat/models/mentions.cljs [hunk 1]
```diff
@@ -100,11 +105,13 @@
...
              (assoc-in [:chat/inputs-with-mentions chat-id] input-segments))}))
 
+(defonce current-call-id (atom 0))
+
 (rf/defn on-change-text
   {:events [:mention/on-change-text]}
   [{:keys [db]} text]
   (let [chat-id (:current-chat-id db)
-        params  [chat-id text]
+        params  [chat-id text (swap! current-call-id inc) (js/Date.now)]
```
Comment (ilmotta) [inline src/legacy/status_im/chat/models/mentions.cljs]: @qfrank, event handlers are the pure data layer in the frontend. We shouldn't deref or mutate atoms.

File: src/legacy/status_im/chat/models/mentions.cljs [hunk 2]
```diff
@@ -118,11 +125,28 @@
...
   (log/debug "[mentions] on-change-text-success" {:result result})
-  (let [{:keys [state chat-id mentionable-users input-segments]} (transfer-mention-result result)]
-    {:db (-> db
-             (assoc-in [:chats/mention-suggestions chat-id] mentionable-users)
-             (assoc-in [:chats/mentions chat-id :mentions] state)
-             (assoc-in [:chat/inputs-with-mentions chat-id] input-segments))}))
+  (let [{:keys [state chat-id mentionable-users
+                input-segments call-id call-time]} (transfer-mention-result result)]
+    (when (= call-id @current-call-id)
+      (debounce/debounce-and-dispatch [:mention/re-render-text-input
```
Comment (ilmotta) [inline src/legacy/status_im/chat/models/mentions.cljs]: Same here, the event handler must return nil or a map of two optional effects `:db` and `:fx`. `debounce-and-dispatch` is a stateful call.

File: src/legacy/status_im/chat/models/mentions.cljs (hunk 1 above)
Comment (ilmotta) [reply src/legacy/status_im/chat/models/mentions.cljs]: The better solution is still simple :) You can store the current call ID in the app-db when `:mention/on-change-text` executes. Then, in the event `:mention/on-change-text-success` simply reads from the db.

The other important change is to use the global interceptor `:now`, this will give you back a unix timestamp in millseconds. The result will be available in the coeffects map. You can destructure like this in the event `{:keys [db now]}`.

Comment (qfrank) [reply src/legacy/status_im/chat/models/mentions.cljs]: fixed

File: src/legacy/status_im/chat/models/mentions.cljs (hunk 2 above)
Comment (qfrank) [reply src/legacy/status_im/chat/models/mentions.cljs]: fixed

File: src/legacy/status_im/chat/models/mentions.cljs [hunk 3]
```diff
@@ -100,29 +103,53 @@
...
+               (assoc-in [:chat/inputs-with-mentions chat-id] input-segments)
+               (assoc-in [:chats/mention-suggestions chat-id] mentionable-users)
+               (assoc-in [:chats/mentions chat-id :mentions] state))
+       :fx [[:dispatch
+             [:mention/re-render-text-input
+              {:chat-id           chat-id
+               :mentionable-users mentionable-users
+               :state             state
+               :input-segments    input-segments
+               :call-id           call-id
```
Comment (ilmotta) [inline src/legacy/status_im/chat/models/mentions.cljs]: `:call-id` and `:response-time` are not being used by the event `:mention/re-render-text-input`.

File: src/legacy/status_im/chat/models/mentions.cljs [hunk 4]
```diff
@@ -100,29 +103,53 @@
...
-    {:db (-> db
-             (assoc-in [:chats/mention-suggestions chat-id] mentionable-users)
-             (assoc-in [:chats/mentions chat-id :mentions] state)
-             (assoc-in [:chat/inputs-with-mentions chat-id] input-segments))}))
+  (let [{:keys [state chat-id mentionable-users
+                input-segments call-id call-time]} (transfer-mention-result result)
+        current-call-id                            (get-in db [:chat/inputs chat-id :current-call-id])]
+    (when (= call-id current-call-id)
+      {:db (-> db
+               (assoc-in [:chat/inputs-with-mentions chat-id] input-segments)
```
Comment (ilmotta) [inline src/legacy/status_im/chat/models/mentions.cljs:129]: Is there a reason that we are storing in the app-db twice with the same data? Once here, and immediately after in `:mention/re-render-text-input`. This seems unnecessary.

File: src/legacy/status_im/chat/models/mentions.cljs [hunk 5]
```diff
@@ -100,29 +103,53 @@
...
+             [:mention/re-render-text-input
+              {:chat-id           chat-id
+               :mentionable-users mentionable-users
+               :state             state
+               :input-segments    input-segments
+               :call-id           call-id
+               :response-time     (- now call-time)}]]]})))
+
+(rf/defn re-render-text-input
+  {:events [:mention/re-render-text-input]}
```
Comment (ilmotta) [inline src/legacy/status_im/chat/models/mentions.cljs]: It's better to use names of events that are not tied to component lifecycles. Re-rendering is a React concept.

File: ios/Podfile.lock [hunk 6]
```diff
@@ -1485,7 +1485,7 @@ SPEC CHECKSUMS:
   FBLazyVector: 56e0e498dbb513b96c40bac6284729ba4e62672d
   FBReactNativeSpec: 146c741a3f40361f6bc13a4ba284678cbedb5881
   fmt: ff9d55029c625d3757ed641535fd4a75fedc7ce9
-  glog: c5d68082e772fa1c511173d6b30a9de2c05a69a2
+  glog: 530710e7949eb12c82670bd09e946becf50a3c2a
```
Comment (ilmotta) [inline ios/Podfile.lock]: I don't use macOS to develop, but is this change really needed?

Comment (qfrank) [reply ios/Podfile.lock]: It seems that every time I run `make run-ios`, this Podfile.lock file will be updated. I generally don't submit this file. However, I was negligent this time. Perhaps it should be added to.gitignore?  cc @siddarthkay

File: src/legacy/status_im/chat/models/mentions.cljs (hunk 5 above)
Comment (qfrank) [reply src/legacy/status_im/chat/models/mentions.cljs]: fixed

File: src/legacy/status_im/chat/models/mentions.cljs (hunk 4 above)
Comment (qfrank) [reply src/legacy/status_im/chat/models/mentions.cljs:129]: fixed

File: src/legacy/status_im/chat/models/mentions.cljs (hunk 3 above)
Comment (qfrank) [reply src/legacy/status_im/chat/models/mentions.cljs]: fixed

File: ios/Podfile.lock (hunk 6 above)
Comment (siddarthkay) [reply ios/Podfile.lock]: Hey @qfrank : the glog hash change is indeed something that should not happen and whenever we install pods we get a different hash from the place we fetch glog.
I logged a similar issue with jitpack for blur dependency -> https://github.com/jitpack/jitpack.io/issues/6494
We should not gitignore the podfile.lock
> Whether or not you check in the Pods directory, the Podfile and Podfile.lock should always be kept under version control.

ref -> https://guides.cocoapods.org/using/using-cocoapods.html#benefits-of-ignoring-the-pods-directory

**Review by ilmotta — APPROVED** [review]

**Comment by churik** [comment]
see https://github.com/status-im/status-mobile/pull/21171#issuecomment-2331566805

**Comment by churik** [comment]
@qfrank not sure why builds are failed, can you have a look? 
Thanks!

**Comment by pavloburykh** [comment]
> @qfrank not sure why builds are failed, can you have a look? Thanks!

@qfrank could you please also rebase corresponding go [PR](https://github.com/status-im/status-go/pull/3806) and update mobile PR as well? Thank you!

**Comment by qfrank** [comment]
Sorry for the delay. I'm just back from holiday. Fixed @churik @pavloburykh

File: src/legacy/status_im/chat/models/mentions.cljs [hunk 7]
```diff
@@ -102,12 +105,14 @@
...
-  [{:keys [db]} text]
+  [{:keys [db now]} text]
   (let [chat-id (:current-chat-id db)
-        params  [chat-id text]
+        call-id (inc (get-in db [:chat/inputs chat-id :current-call-id] 0))
+        params  [chat-id text call-id now]
         method  "wakuext_chatMentionOnChangeText"]
     (log/debug "[mentions] on-change-text" {:params params})
-    {:json-rpc/call [{:method     method
+    {:db            (assoc-in db [:chat/inputs chat-id :current-call-id] call-id)
```
Comment (flexsurfer) [inline src/legacy/status_im/chat/models/mentions.cljs]: probably better `current-mentions-call-id`

File: src/legacy/status_im/chat/models/mentions.cljs [hunk 8]
```diff
@@ -116,10 +121,18 @@
...
-  [{:keys [db]} result]
+  [{:keys [db now]} result]
   (log/debug "[mentions] on-change-text-success" {:result result})
-  (let [{:keys [chat-id mentionable-users]} (transfer-mention-result result)]
-    {:db (assoc-in db [:chats/mention-suggestions chat-id] mentionable-users)}))
+  (let [{:keys [state chat-id mentionable-users
+                input-segments call-id call-time]} (transfer-mention-result result)
+        current-call-id                            (get-in db [:chat/inputs chat-id :current-call-id])]
+    (when (= call-id current-call-id)
+      (let [response-time (- now call-time)]
```
Comment (flexsurfer) [inline src/legacy/status_im/chat/models/mentions.cljs]: why call time is needed? if it's needed only for debug I guess it's better to remove it

**Comment by pavloburykh** [comment]
@qfrank thank you for the PR. No issues from my side. Ready for merge.

File: src/legacy/status_im/chat/models/mentions.cljs (hunk 8 above)
Comment (qfrank) [reply src/legacy/status_im/chat/models/mentions.cljs]: removed

File: src/legacy/status_im/chat/models/mentions.cljs (hunk 7 above)
Comment (qfrank) [reply src/legacy/status_im/chat/models/mentions.cljs]: renamed

**Review by flexsurfer — APPROVED** [review]
